SQOOP-3438 Sqoop Import with create hcatalog table for ORC will not w…

dionusos opened the pull request:
…ork with Hive 2 as the table created would be a ACID table and transactional

Change-Id: Ic6a5600fbc351c98dd29ec5809ac99680799abc0

Commit:
SQOOP-3438 Sqoop Import with create hcatalog table for ORC will not work in HDP 3.x as the table created would be a ACID table and transactional

Change-Id: Ic6a5600fbc351c98dd29ec5809ac99680799abc0

Commit:
SQOOP-3438 Sqoop Import with create hcatalog table for ORC will not work with Hive3 as the table created would be a ACID table and transactional
Adding tests

Change-Id: I03eb99c469e030deff84982c724a72f311a0f631

dvoros commented on src/java/org/apache/sqoop/SqoopOptions.java:
```diff
@@ -1654,6 +1656,14 @@ public String getHCatTableName() {
     return this.hCatTableName;
   }
 
+  public void useExternalHCatTable(boolean value) {
```
Nit: Other boolean setter methods are called `set*` here.

dionusos replied:
Thanks @dvoros

dvoros commented on src/docs/user/hcatalog.txt:
```diff
@@ -61,6 +61,10 @@ The presence of the +--hcatalog-table+ option signifies that the import
 HCatalog jobs.
 
++--hcatalog-external-table+::
+Set this flag if you need to create external Hive table for example to store
```
Instead of ORC, I'd mention non-transactional tables, since that's stopping you from creating HCat table without this patch.

dvoros reviewed (commented):
👍 + left some comments.

ebogi commented on src/docs/user/hcatalog.txt:
```diff
@@ -61,6 +61,10 @@ The presence of the +--hcatalog-table+ option signifies that the import
 or export job is done using HCatalog tables, and it is a required option for
 HCatalog jobs.
 
++--hcatalog-external-table+::
```
Could you please extend the documentation of --drop-and-create-hcatalog-table option that it will have affect on external HCat tables too?

dionusos replied:
You are correct, thanks for notifying me about the use case with delete. I think the provided example fits there.

ebogi commented on src/test/org/apache/sqoop/hcat/HCatalogImportTest.java:
```diff
@@ -770,6 +770,29 @@ public void testTableCreation() throws Exception {
   }
 
+  @Test
+  public void testExternalTableCreation() throws Exception {
```
Could you please add more test cases? E.g. for when the table already exists and either exception is thrown or we drop it based on the usage of --drop-and-create-hcatalog-table option?

dionusos replied:
One test case for drop and create and one negative is added.

ebogi replied:
Thanks for adding new test cases!

ebogi commented on src/docs/user/hcatalog.txt:
```diff
@@ -61,6 +61,10 @@ The presence of the +--hcatalog-table+ option signifies that the import
 or export job is done using HCatalog tables, and it is a required option for
 HCatalog jobs.
 
++--hcatalog-external-table+::
```
Using this option makes sense only together with --create-hcatalog-table thus could you please add fail fast scenarios to prevent separate usage? Also could you please update the documentation accordingly?
For reference please see SQOOP-816 and SQOOP-3324 where external Hive table option has been introduced and documented.

ebogi replied:
Thanks for extending the documentation!

ebogi reviewed (changes requested):
Thanks Denes for adding this new capability to Sqoop! I think your change generally looks good just need to cover some edge cases and need to have some more testing and documentation. Please find my comments inline.

Commit:
SQOOP-3438 Sqoop Import with create hcatalog table for ORC will not work with Hive3 as the table created would be a ACID table and transactional
 - Adding negative test case
 - Extending documentation
 - Parameter / option checking

Change-Id: Id2830c3cecc78d77cb83a7f18dc7e2e7b14850e3

ebogi commented on src/test/org/apache/sqoop/hcat/HCatalogImportTest.java:
```diff
@@ -770,6 +771,80 @@ public void testTableCreation() throws Exception {
+  }
+
+  @Test
+  public void testExternalTableDropAndCreation() throws Exception {
```
This test case currently tests only that --drop-and-create-hcatalog-table and --hcatalog-external-tabl can present together. For testing the drop-if-exist behavior there could be two test cases added: pre-create an external table (e.g. via importing one) and then try to import the same table
1. with --drop-and-create-hcatalog-table --> should succeed
2. without --drop-and-create-hcatalog-table --> should throw an exception

ebogi commented on src/test/org/apache/sqoop/hcat/HCatalogImportTest.java:
```diff
@@ -770,6 +771,80 @@ public void testTableCreation() throws Exception {
+    addlArgsArray.add("\"stored as orc tblproperties (\"transactional\"=\"false\")\"");
+    setExtraArgs(addlArgsArray);
+    utils.dropHCatTableIfExists(table, SqoopHCatUtilities.DEFHCATDB);
+    try {
```
Could you please use ExpectedException rule instead of using a try-catch block here?

dionusos replied:
Of course, that is a much cleaner solution. I did follow the pattern of the original code.

ebogi replied:
Boy scout rule here again, thanks for updating it!

ebogi commented on src/test/org/apache/sqoop/hcat/HCatalogImportTest.java:
```diff
@@ -770,6 +771,80 @@ public void testTableCreation() throws Exception {
+
+  @Test
+  public void testExternalTableDropAndCreation() throws Exception {
+    final int TOTAL_RECORDS = 1 * 10;
```
This and the next ~15 lines present in all the new test cases. Could you please extract these into separate methods to eliminate code duplication?

dionusos replied:
As I saw into the other test cases I found that code duplication is the pattern here. I am glad that I was wrong.

ebogi replied:
Yeah, we generally try to follow the boy scout rule meaning some minimal refactor of legacy code if it results in more clean code.

ebogi reviewed (changes requested):
Thank you very much for the updates!
I have some more findings that should be addressed before merging your change, please find them below.

Commit:
SQOOP-3438 Sqoop Import with create hcatalog table for ORC will not work with Hive3 as the table created would be a ACID table and transactional
- refactor test cases

Change-Id: I0998227351ebd049cb2f5c485ee500effaab5b31

Commit:
SQOOP-3438 Sqoop Import with create hcatalog table for ORC will not work with Hive3 as the table created would be a ACID table and transactional
- implemented reqested test changes
- removed unnecessary refactoring in old code

Change-Id: I8ac7e9b10669615fa479e13e02d27b877bdad462

ebogi reviewed (approved):
Thanks Denes for improving your patch! It LGTM now, ship it!
